src/utils/Switch.ts

```diff
@@ -1,94 +1,221 @@
+}
+
+export interface ICaseMatchWithoutInput<V> {
+    case<R>(condition:boolean, resultToUseIfMatch:R):ISwitchWithoutInput<V | R>
+    caseLazy<R>(condition:boolean, resultProviderToUseIfMatch:Provider<R>):ISwitchWithoutInput<V | R>
+    matches<R>(predicate:Provider<boolean>, resultToUseIfMatch:R): ISwitchWithoutInput<V | R>
+    matchesLazy<R>(predicate:Provider<boolean>, resultProviderToUseIfMatch:Provider<R>): ISwitchWithoutInput<V | R>
+}
+
+export interface ISwitchResult<V> {
+    result:Maybe<V>;
+}
+
+export interface ISwitchWithoutInput<V> extends ICaseMatchWithoutInput<V>, ISwitchResult<V> {}
+
+const SwitchWithExplicitInput = <T>(input:T):ISwitchWithEarlyInput<T, never> => {
+    return SwitchWithExplicitInputImpl(input, [])
+}
+
+type ExplicitInputPredicateMapperPair<T, R> = Pair<Predicate<T>, Mapper<T, R>>
+type ExplicitInputPredicateMapperPairs<T, R> = Array<ExplicitInputPredicateMapperPair<T, R>>
+
+const SwitchWithExplicitInputImpl = <T, V>(input:T, predicateMapperPairs:ExplicitInputPredicateMapperPairs<T, V>):ISwitchWithEarlyInput<T, V> => ({
+    case<R>(possiblyEqualValue:T, resultToUseIfMatch:R):ISwitchWithEarlyInput<T, R | V> {
+        return this.matchesLazy(
+            equals(possiblyEqualValue),
+            Constant(resultToUseIfMatch)
+        )
+    },
+    cases<R>(possiblyEqualValues:T[], mapperToUseIfMatch:Mapper<T, R>):ISwitchWithEarlyInput<T, V | R> {
+        return this.matchesLazy(
+            item => possiblyEqualValues.includes(item),
+            mapperToUseIfMatch
+        )
+    },
+    caseLazy<R>(possiblyEqualValue:T, mapperToUseIfMatch:Mapper<T, R>):ISwitchWithEarlyInput<T, V | R> {
+        return this.matchesLazy(
+            equals(possiblyEqualValue),
+            mapperToUseIfMatch
+        )
+    },
+    matches<R>(predicate:Predicate<T>, resultToUseIfMatch:R): ISwitchWithEarlyInput<T, V | R> {
+        return this.matchesLazy(
+            predicate,
+            Constant(resultToUseIfMatch)
+        )
+    },
+    matchesLazy<R>(predicate:Predicate<T>, resultMapperToUseIfMatch:Mapper<T, R>): ISwitchWithEarlyInput<T, R | V> {
+        return SwitchWithExplicitInputImpl<T, R | V>(
+            input, 
+            [
+                ...predicateMapperPairs, 
+                [predicate, resultMapperToUseIfMatch]
+            ]
+        )
+    },
+    get result():Maybe<V> {
+        return first(
+            predicateMapperPairs,
+            ([predicate]) => predicate(input)
+        ).map(
+            ([_, mapper]) => mapper(input)
+        )
+    },
+})
 
-export const EmptySwitch = ():ISwitch<never,never> => ({
-    case<S, R>(possiblyEqualValue:S, resultToUseIfMatch:R):ISwitch<S, R> {
-        return this.matchesLazy(input => input == possiblyEqualValue, () => resultToUseIfMatch)
+const SwitchWithoutInput = ():ISwitchWithoutInput<never> => EmptySwitchWithoutInput();
```

Comment:
this function is unused

Reply:
Not even in the tests? I’ve wanted this in the past, for when you’re like “if a > 5” do x; if b < 7, do y” where the predicates don’t depend on any one input so I wanted to add that capability before I need it

Reply:
Not even in the tests? I’ve wanted this in the past, for when you’re like “if a > 5” do x; if b < 7, do y” where the predicates don’t depend on any one input so I wanted to add that capability before I need it

Reply:
well, you could export it then! but correct, there are no tests around this one either. It's greyed out in the IDE

Reply:
thanks mike! deleted it, thought it was something else earlier! :)